relay-profiling/src/sample/v2.rs

```diff
@@ -0,0 +1,159 @@
+    stacks: Vec<Vec<usize>>,
+    frames: Vec<Frame>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    thread_metadata: Option<BTreeMap<String, ThreadMetadata>>,
+}
+
+impl ProfileData {
+    /// Ensures valid profile chunk or returns an error.
+    ///
+    /// Mutates the profile chunk. Removes invalid samples and threads.
+    /// Throws an error if the profile chunk is malformed.
+    /// Removes extra metadata that are not referenced in the samples.
+    ///
+    /// profile.normalize("cocoa", "arm64e")
+    pub fn normalize(&mut self, platform: &str) -> Result<(), ProfileError> {
+        if self.samples.is_empty() {
+            return Err(ProfileError::NotEnoughSamples);
+        }
+
+        if !self.all_stacks_referenced_by_samples_exist() {
+            return Err(ProfileError::MalformedSamples);
+        }
+
+        if !self.all_frames_referenced_by_stacks_exist() {
+            return Err(ProfileError::MalformedStacks);
+        }
+
+        self.strip_pointer_authentication_code(platform);
+        self.remove_unreferenced_threads();
+
+        Ok(())
+    }
+
+    fn strip_pointer_authentication_code(&mut self, platform: &str) {
+        let addr = match platform {
+            // https://github.com/microsoft/plcrashreporter/blob/748087386cfc517936315c107f722b146b0ad1ab/Source/PLCrashAsyncThread_arm.c#L84
+            "cocoa" => 0x0000000FFFFFFFFF,
+            _ => return,
+        };
+        for frame in &mut self.frames {
+            frame.strip_pointer_authentication_code(addr);
+        }
+    }
+
+    /// Checks that all stacks referenced by the samples exist in the stacks.
+    fn all_stacks_referenced_by_samples_exist(&self) -> bool {
+        self.samples
+            .iter()
+            .all(|sample| self.stacks.get(sample.stack_id).is_some())
+    }
+
+    /// Checks that all frames referenced by the stacks exist in the frames.
+    fn all_frames_referenced_by_stacks_exist(&self) -> bool {
+        self.stacks.iter().all(|stack| {
+            stack
+                .iter()
+                .all(|frame_id| self.frames.get(*frame_id).is_some())
+        })
+    }
+
+    fn remove_unreferenced_threads(&mut self) {
+        if let Some(thread_metadata) = &mut self.thread_metadata {
+            let thread_ids = self
+                .samples
+                .iter()
+                .map(|sample| sample.thread_id.clone())
+                .collect::<HashSet<_>>();
+            thread_metadata.retain(|thread_id, _| thread_ids.contains(thread_id));
+        }
+    }
+}
+
+fn parse_profile(payload: &[u8]) -> Result<ProfileChunk, ProfileError> {
+    let d = &mut serde_json::Deserializer::from_slice(payload);
+    let mut profile: ProfileChunk =
+        serde_path_to_error::deserialize(d).map_err(ProfileError::InvalidJson)?;
+
+    profile
+        .profile
+        .normalize(profile.metadata.platform.as_str())?;
+
+    Ok(profile)
+}
+
+pub fn parse(payload: &[u8]) -> Result<Vec<u8>, ProfileError> {
```

Comment:
Docs would be good here, not sure why the function is that way in the first place, the caller could just use `parse_profile` and then serialize?

Is it supposed to normalize it -> better name?

Reply:
It's coming from other formats: we needed to modify the profile after it's parsed and normalized for some reasons, before we returned the serialized payload. It was also useful to separate those 2 functions for tests as I wanted to check specific values of the field and getting back a serialized payload was unhelpful.

I'll see what I can do here, it's likely possible to compress into one function.

Reply:
Feel free to leave it and just give the function a more descriptive name, maybe just `parse_and_normalize` and a quick doc string?

Reply:
I refactored it the way you suggested.